pkg/koordlet/util/perf/perf_linux.go

```diff
@@ -22,15 +22,377 @@ package perf
+				pool := make([]byte, 24+eventNum*16)
+				return &pool
+			},
+		}
+	}
+}
+
+func LibInit() {
+	perfValuePool.New = func() interface{} {
+		return &value{}
+	}
+	initlibpfm.Do(func() {
+		if err := C.pfm_initialize(); err != C.PFM_SUCCESS {
+			klog.Fatalf("Unable to init libpfm: %v", err)
+		}
+	})
+}
+
+func LibFinalize() {
+	closelibpfm.Do(func() {
+		C.pfm_terminate()
+	})
+}
+
+type PerfGroupCollector struct {
+	cgroupFile     *os.File
+	cpus           []int
+	perfCollectors sync.Map
+	idEventMap     map[uint64]string
+	mu             sync.Mutex
+	resultMap      map[string]float64
+	valueCh        chan perfValue
+	syscall6       func(trap uintptr, a1 uintptr, a2 uintptr, a3 uintptr, a4 uintptr, a5 uintptr, a6 uintptr) (r1 uintptr, r2 uintptr, err syscall.Errno)
+	closeCh        chan struct{}
+}
+
+type perfCollector struct {
+	cpu      int
+	syscall6 func(trap uintptr, a1 uintptr, a2 uintptr, a3 uintptr, a4 uintptr, a5 uintptr, a6 uintptr) (r1 uintptr, r2 uintptr, err syscall.Errno)
+	leaderFd io.ReadCloser
+	fds      []io.ReadCloser
+}
+
+type perfValue struct {
+	Value float64
+	ID    uint64
+}
+
+type value struct {
+	Value uint64
+	ID    uint64
+}
+
+type perfValueHeader struct {
+	Nr          uint64 // number of events
+	TimeEnabled uint64 // time event active
+	TimeRunning uint64 // time event on CPU
+}
+
+// first event is group leader
+func NewPerfGroupCollector(cgroupFile *os.File, cpus []int, events []string, syscallFunc func(trap, a1, a2, a3, a4, a5, a6 uintptr) (r1, r2 uintptr, err syscall.Errno)) (collector *PerfGroupCollector, err error) {
+	if len(events) == 0 {
+		err = errors.New("events cannot be empty")
+		return nil, err
+	}
+	collector = &PerfGroupCollector{
+		cgroupFile:     cgroupFile,
+		cpus:           cpus,
+		perfCollectors: sync.Map{},
+		idEventMap:     make(map[uint64]string),
+		mu:             sync.Mutex{},
+		resultMap:      make(map[string]float64),
+		valueCh:        make(chan perfValue),
+		syscall6:       syscallFunc,
+		closeCh:        make(chan struct{}),
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(cpus))
+	attrMap := make(map[string]*unix.PerfEventAttr, len(events))
+	for i, event := range events {
+		if attr, e := createPerfConfig(event); e != nil {
+			err = e
+			return nil, err
+		} else {
+			// first event is group leader
+			if i == 0 {
+				attr.Bits |= unix.PerfBitDisabled
+			}
+			attr.Read_format = unix.PERF_FORMAT_GROUP | unix.PERF_FORMAT_TOTAL_TIME_ENABLED | unix.PERF_FORMAT_TOTAL_TIME_RUNNING | unix.PERF_FORMAT_ID
+			attr.Sample_type = unix.PERF_SAMPLE_IDENTIFIER
+			attr.Size = uint32(unsafe.Sizeof(unix.PerfEventAttr{}))
+			attr.Bits |= unix.PerfBitInherit
+			attrMap[event] = attr
+		}
+	}
+	var errMutex sync.Mutex
+	for _, cpu := range cpus {
+		go func(cpu int) {
+			// create perf group
+			var pc perfCollector
+			pc.syscall6 = syscallFunc
+			pc.fds = make([]io.ReadCloser, 0, len(events)-1)
+			pc.cpu = cpu
+			defer wg.Done()
+			attr := attrMap[events[0]]
```

Comment:
what is the difference between events 0 and the others, maybe use different functions for cycles and instructions?

Reply:
events[0] is leader event. Some perf attr bit is different.